cuda/solver/batch_idr_kernels.cu

```diff
@@ -83,64 +84,46 @@ namespace batch_idr {
 template <typename T>
 using BatchIdrOptions = gko::kernels::batch_idr::BatchIdrOptions<T>;
 
-#define BATCH_IDR_KERNEL_LAUNCH(_stoppertype, _prectype)                   \
-    apply_kernel<stop::_stoppertype<ValueType>>                            \
-        <<<nbatch, default_block_size, shared_size>>>(                     \
-            opts.max_its, opts.residual_tol, opts.subspace_dim_val,        \
-            opts.kappa_val, opts.to_use_smoothing, opts.deterministic_gen, \
-            logger, _prectype<ValueType>(), subspace_vectors, a, b.values, \
-            x.values)
-
-template <typename BatchMatrixType, typename LogType, typename ValueType>
-static void apply_impl(std::shared_ptr<const CudaExecutor> exec,
-                       const BatchIdrOptions<remove_complex<ValueType>> opts,
-                       LogType logger, const BatchMatrixType& a,
-                       const gko::batch_dense::UniformBatch<const ValueType>& b,
-                       const gko::batch_dense::UniformBatch<ValueType>& x,
-                       const ValueType* const subspace_vectors)
-{
-    using real_type = gko::remove_complex<ValueType>;
-    const size_type nbatch = a.num_batch;
-    static_assert(default_block_size >= 2 * config::warp_size,
-                  "Need at least two warps!");
-
-    int shared_size =
-#if GKO_CUDA_BATCH_USE_DYNAMIC_SHARED_MEM
-        gko::kernels::batch_idr::local_memory_requirement<ValueType>(
-            a.num_rows, b.num_rhs, opts.subspace_dim_val);
-#else
-        0;
-#endif
-
-    if (opts.preconditioner == gko::preconditioner::batch::type::none) {
-#if GKO_CUDA_BATCH_USE_DYNAMIC_SHARED_MEM
-        shared_size +=
-            BatchIdentity<ValueType>::dynamic_work_size(a.num_rows, a.num_nnz) *
-            sizeof(ValueType);
-#endif
-        if (opts.tol_type == gko::stop::batch::ToleranceType::absolute) {
-            BATCH_IDR_KERNEL_LAUNCH(SimpleAbsResidual, BatchIdentity);
-        } else {
-            BATCH_IDR_KERNEL_LAUNCH(SimpleRelResidual, BatchIdentity);
-        }
-    } else if (opts.preconditioner ==
-               gko::preconditioner::batch::type::jacobi) {
-#if GKO_CUDA_BATCH_USE_DYNAMIC_SHARED_MEM
-        shared_size +=
-            BatchJacobi<ValueType>::dynamic_work_size(a.num_rows, a.num_nnz) *
-            sizeof(ValueType);
-#endif
-
-        if (opts.tol_type == gko::stop::batch::ToleranceType::absolute) {
-            BATCH_IDR_KERNEL_LAUNCH(SimpleAbsResidual, BatchJacobi);
-        } else {
-            BATCH_IDR_KERNEL_LAUNCH(SimpleRelResidual, BatchJacobi);
-        }
-    } else {
-        GKO_NOT_IMPLEMENTED;
+template <typename CuValueType>
+class KernelCaller {
```

Comment:
Instead of declaring this class in every solver file, you could declare it in a common (across solvers and back-ends) file with some default implementations if possible and override the method in each solver and back-end.

Maybe some Macro magic (replacing Solver names) and variadic macros will help reducing a lot of this code duplication,

Reply:
In general, different solvers might also need different constructors for the `KernelCaller`. See the cuda batch IDR, for example. Different solvers would have different shared shared memory logic and stuff too. So we would not be saving much, and it's not worth introducing macros for this. In the future, we can see if more things can be abstracted into some kind of parent class, at least for each solver across all backends.